File: test/drivingReducer.test.ts
// Pure-function tests for the Driving state machine — no React renderer
// needed. Each case names the event being delivered, the expected next
// state, and the side-effect intents the reducer should emit.

import { describe, it, expect } from 'vitest';
import {
  initialContext,
  reduce,
  type DrivingContext,
} from '../client/src/voice/drivingReducer';
import {
  composeTranscript,
  displayedCaptionText,
  isCurrentTurnTranscribing,
} from '../client/src/voice/drivingLoop';

const idle: DrivingContext = { ...initialContext };
const recording: DrivingContext = { ...initialContext, state: 'recording' };
const thinking: DrivingContext = {
  ...initialContext,
  state: 'thinking',
  lastUserText: 'hello',
};
const ai: DrivingContext = {
  ...initialContext,
  state: 'ai',
  lastUserText: 'hello',
  lastReplyText: 'hi there',
  liveReplyText: 'hi there',
};

describe('idle', () => {
  it('tap moves to recording and arms the mic', () => {
    const { next, side } = reduce(idle, { type: 'tap' });
    expect(next.state).toBe('recording');
    expect(next.error).toBeNull();
    expect(side).toEqual([{ kind: 'startMic' }]);
  });

  it('ignores unrelated events', () => {
    expect(reduce(idle, { type: 'silence' }).next.state).toBe('idle');
    expect(reduce(idle, { type: 'stt.done', text: 'x' }).next.state).toBe('idle');
    expect(reduce(idle, { type: 'tts.done' }).next.state).toBe('idle');
  });
});

describe('recording', () => {
  it('tap stops the mic and enters thinking', () => {
    const { next, side } = reduce(recording, { type: 'tap' });
    expect(next.state).toBe('thinking');
    expect(side).toEqual([{ kind: 'stopMic' }]);
  });

  it('stt.error cancels back to idle with the reason surfaced', () => {
    const { next, side } = reduce(recording, {
      type: 'stt.error',
      reason: 'mic_denied',
    });
    expect(next.state).toBe('idle');
    expect(next.error).toBe('mic_denied');
    expect(side).toEqual([{ kind: 'cancelMic' }]);
  });
});

describe('thinking', () => {
  it('reply.done stays thinking, hides the reply text, and arms the TTS player', () => {
    const { next, side } = reduce(thinking, {
      type: 'reply.done',
      text: 'sup',
    });
    expect(next.state).toBe('thinking');
    expect(next.pendingReplyText).toBe('sup');
    expect(next.lastReplyText).toBe('');
    expect(next.liveReplyText).toBe('');
    expect(side).toEqual([{ kind: 'armTts' }]);
  });

  it('tts.start with a pending reply reveals the response and enters ai', () => {
    const { next, side } = reduce(
      { ...thinking, pendingReplyText: 'sup', lastReplyText: 'previous audible reply' },
      { type: 'tts.start' },
    );

    expect(next.state).toBe('ai');
    expect(next.pendingReplyText).toBe('');
    expect(next.lastReplyText).toBe('sup');
    expect(next.liveReplyText).toBe('sup');
    expect(side).toEqual([]);
  });

  it('stale tts.start without a pending reply does not show previous turn text', () => {
    const { next, side } = reduce(
      {
        ...thinking,
        pendingReplyText: '',
        lastReplyText: 'previous audible reply',
        liveReplyText: '',
      },
      { type: 'tts.start' },
    );

    expect(next.state).toBe('thinking');
    expect(next.pendingReplyText).toBe('');
    expect(next.lastReplyText).toBe('previous audible reply');
    expect(next.liveReplyText).toBe('');
    expect(side).toEqual([]);
  });

  it('reply.done then tts.error returns idle without exposing the pending reply', () => {
    let ctx = reduce(thinking, { type: 'reply.done', text: 'unheard reply' }).next;

    const { next, side } = reduce(ctx, {
      type: 'tts.error',
      reason: 'openclaw_infer_tts_failed',
    });

    expect(next.state).toBe('idle');
    expect(next.error).toBe('openclaw_infer_tts_failed');
    expect(next.pendingReplyText).toBe('');
    expect(next.lastReplyText).toBe('');
    expect(next.liveReplyText).toBe('');
    expect(side).toEqual([]);
  });

  it('reply.done then tts.done returns idle without exposing the pending reply', () => {
    const ctx = reduce(thinking, { type: 'reply.done', text: 'unheard reply' }).next;

    const { next, side } = reduce(ctx, { type: 'tts.done' });

    expect(next.state).toBe('idle');
    expect(next.pendingReplyText).toBe('');
    expect(next.lastReplyText).toBe('');
    expect(next.liveReplyText).toBe('');
    expect(side).toEqual([]);
  });

  it('reply.error returns to idle with the reason', () => {
    const { next, side } = reduce({ ...thinking, pendingReplyText: 'unheard reply' }, {
      type: 'reply.error',
      reason: 'xai_http_500',
    });
    expect(next.state).toBe('idle');
    expect(next.error).toBe('xai_http_500');
    expect(next.pendingReplyText).toBe('');
    expect(next.lastReplyText).toBe('');
    expect(next.liveReplyText).toBe('');
    expect(side).toEqual([]);
  });

  it('reply.done then tap cancel clears the pending unseen reply', () => {
    const ctx = reduce(thinking, { type: 'reply.done', text: 'unheard reply' }).next;

    const { next, side } = reduce(ctx, { type: 'tap' });

    expect(next.state).toBe('idle');
    expect(next.pendingReplyText).toBe('');
    expect(next.lastReplyText).toBe('');
    expect(next.liveReplyText).toBe('');
    expect(side).toEqual([{ kind: 'cancelReply' }]);
  });

  it('tap during thinking keeps waiting while current turn STT is still finalizing', () => {
    const { next, side } = reduce(thinking, { type: 'tap', currentTurnTranscribing: true });
    expect(next.state).toBe('thinking');
    expect(side).toEqual([]);
  });

  it('stt.done just records the user text without state change', () => {
    const { next, side } = reduce(thinking, { type: 'stt.done', text: 'HELLO' });
    expect(next.state).toBe('thinking');
    expect(next.lastUserText).toBe('HELLO');
    expect(side).toEqual([]);
  });
});

describe('ai', () => {
  it('tts.done returns to idle with no further side-effects', () => {
    const { next, side } = reduce(ai, { type: 'tts.done' });
    expect(next.state).toBe('idle');
    expect(side).toEqual([]);
  });

  it('silence stops TTS and returns to idle', () => {
    const { next, side } = reduce(ai, { type: 'silence' });
    expect(next.state).toBe('idle');
    expect(side).toEqual([{ kind: 'stopTts' }]);
  });

  it('tap stops TTS and returns to idle', () => {
    const { next, side } = reduce(ai, { type: 'tap' });
    expect(next.state).toBe('idle');
    expect(side).toEqual([{ kind: 'stopTts' }]);
  });

  it('tts.error surfaces the reason and returns to idle', () => {
    const { next, side } = reduce(ai, { type: 'tts.error', reason: 'x' });
    expect(next.state).toBe('idle');
    expect(next.error).toBe('x');
    expect(side).toEqual([]);
  });
});

describe('full happy-path sequence', () => {
  it('idle → recording → thinking → reply ready hidden → ai → idle', () => {
    let ctx = idle;
    ctx = reduce(ctx, { type: 'tap' }).next;
    expect(ctx.state).toBe('recording');
    ctx = reduce(ctx, { type: 'tap' }).next;
    expect(ctx.state).toBe('thinking');
    ctx = reduce(ctx, { type: 'stt.done', text: 'hi' }).next;
    expect(ctx.state).toBe('thinking');
    ctx = reduce(ctx, { type: 'reply.done', text: 'hello' }).next;
    expect(ctx.state).toBe('thinking');
    expect(ctx.pendingReplyText).toBe('hello');
    expect(ctx.liveReplyText).toBe('');
    ctx = reduce(ctx, { type: 'tts.start' }).next;
    expect(ctx.state).toBe('ai');
    expect(ctx.pendingReplyText).toBe('');
    expect(ctx.liveReplyText).toBe('hello');
    ctx = reduce(ctx, { type: 'tts.done' }).next;
    expect(ctx.state).toBe('idle');
    expect(ctx.lastReplyText).toBe('hello');
    expect(ctx.lastUserText).toBe('hi');
  });
});


describe('session snapshot replay', () => {
  it('hydrates a completed reconnect snapshot after missed reply/tts events from a fresh idle reducer', () => {
    const { next, side } = reduce(idle, {
      type: 'session.replay',
      events: [
        { type: 'reply.done', text: 'Pull over safely.' },
        { type: 'tts.start' },
        { type: 'tts.done' },
      ],
      hydration: {
        context: {
          ...initialContext,
          state: 'idle',
          lastUserText: 'What should I do?',
          lastReplyText: 'Pull over safely.',
        },
      },
    });

    expect(next.state).toBe('idle');
    expect(next.lastUserText).toBe('What should I do?');
    expect(next.lastReplyText).toBe('Pull over safely.');
    expect(next.pendingReplyText).toBe('');
    expect(next.liveReplyText).toBe('');
    expect(side).toEqual([]);
  });

  it('hydrates a reply-ready reconnect snapshot and arms TTS even when replayed events were ignored from idle', () => {
    const { next, side } = reduce(idle, {
      type: 'session.replay',
      events: [{ type: 'reply.done', text: 'One moment.' }],
      hydration: {
        context: {
          ...initialContext,
          state: 'thinking',
          lastUserText: 'Are you there?',
          pendingReplyText: 'One moment.',
        },
        armTts: true,
      },
    });

    expect(next.state).toBe('thinking');
    expect(next.lastUserText).toBe('Are you there?');
    expect(next.pendingReplyText).toBe('One moment.');
    expect(next.lastReplyText).toBe('');
    expect(next.liveReplyText).toBe('');
    expect(side).toEqual([{ kind: 'armTts' }]);
  });

  it('treats missing disconnectedMs as auto-resumable for older snapshot payloads', () => {
    const { next, side } = reduce(idle, {
      type: 'session.replay',
      events: [{ type: 'reply.done', text: 'spoken reply' }],
      hydration: {
        context: {
          ...initialContext,
          state: 'thinking',
          lastUserText: 'hello',
          pendingReplyText: 'spoken reply',
        },
        armTts: true,
      },
    });

    expect(next.state).toBe('thinking');
    expect(next.lastUserText).toBe('hello');
    expect(next.pendingReplyText).toBe('spoken reply');
    expect(next.liveReplyText).toBe('');
    expect(side).toEqual([{ kind: 'armTts' }]);
  });

  it('hydrates a stale reply-ready reconnect as idle replayable text without arming TTS', () => {
    const { next, side } = reduce(idle, {
      type: 'session.replay',
      events: [{ type: 'reply.done', text: 'One moment.' }],
      hydration: {
        context: {
          ...initialContext,
          state: 'idle',
          lastUserText: 'hello',
          lastReplyText: 'spoken reply',
        },
        armTts: false,
      },
    });

    expect(next.state).toBe('idle');
    expect(next.lastUserText).toBe('hello');
    expect(next.lastReplyText).toBe('spoken reply');
    expect(next.pendingReplyText).toBe('');
    expect(next.liveReplyText).toBe('');
    expect(side).toEqual([]);
  });

  it('hydrates a stale speaking reconnect as idle replayable text without active AI playback', () => {
    const { next, side } = reduce(idle, {
      type: 'session.replay',
      events: [{ type: 'tts.start' }],
      hydration: {
        context: {
          ...initialContext,
          state: 'idle',
          lastUserText: 'hello',
          lastReplyText: 'spoken reply',
        },
        armTts: false,
      },
    });

    expect(next.state).toBe('idle');
    expect(next.lastUserText).toBe('hello');
    expect(next.lastReplyText).toBe('spoken reply');
    expect(next.liveReplyText).toBe('');
    expect(side).toEqual([]);
  });

});

describe('displayedCaptionText', () => {
  it('keeps the live transcript visible while thinking before final STT arrives', () => {
    expect(displayedCaptionText({ ...initialContext, state: 'thinking' }, 'Okay, how we doing?')).toBe(
      'Okay, how we doing?',
    );
  });

  it('does not reuse a previous turn transcript while a new turn is transcribing', () => {
    expect(
      displayedCaptionText(
        { ...initialContext, state: 'thinking', lastUserText: 'Oh my fucking god' },
        '',
      ),
    ).toBe('');
  });

  it('keeps the current final user transcript visible while waiting for the AI reply', () => {
    expect(
      displayedCaptionText(
        { ...initialContext, state: 'thinking', lastUserText: 'previous turn' },
        'current turn',
      ),
    ).toBe('current turn');
  });

  it('switches to the AI reply text only once the AI state starts', () => {
    expect(
      displayedCaptionText(
        { ...initialContext, state: 'ai', lastUserText: 'user words', liveReplyText: 'ai words' },
        'user words',
      ),
    ).toBe('ai words');
  });

  it('does not reveal a pending reply while still thinking', () => {
    expect(
      displayedCaptionText(
        {
          ...initialContext,
          state: 'thinking',
          lastUserText: 'user words',
          pendingReplyText: 'hidden ai words',
          liveReplyText: '',
        },
        'user words',
      ),
    ).toBe('user words');
  });
});

describe('isCurrentTurnTranscribing', () => {
  it('stays true while a current thinking turn is waiting for stt.done', () => {
    expect(
      isCurrentTurnTranscribing('thinking', {
        active: true,
        sttDone: false,
      }),
    ).toBe(true);
  });

  it('turns false after current stt.done', () => {
    expect(
      isCurrentTurnTranscribing('thinking', {
        active: true,
        sttDone: true,
      }),
    ).toBe(false);
  });

  it('does not infer transcribing from previous-turn text', () => {
    expect(
      isCurrentTurnTranscribing('thinking', {
        active: false,
        sttDone: false,
      }),
    ).toBe(false);
  });
});

describe('composeTranscript', () => {
  it('appends the current partial to committed final segments instead of replacing them', () => {
    expect(composeTranscript(['Okay, how we doing?', 'I really hope'], 'Oh my fucking god')).toBe(
      'Okay, how we doing? I really hope Oh my fucking god',
    );
  });
});

File: android/app/src/test/java/app/clawkietalkie/DrivingReducerTest.kt
package app.clawkietalkie

import app.clawkietalkie.voice.CurrentTurnTranscript
import app.clawkietalkie.voice.DrivingContext
import app.clawkietalkie.voice.DrivingEvent
import app.clawkietalkie.voice.DrivingSideEffect
import app.clawkietalkie.voice.DrivingState
import app.clawkietalkie.voice.composeTranscript
import app.clawkietalkie.voice.displayedCaptionText
import app.clawkietalkie.voice.initialDrivingContext
import app.clawkietalkie.voice.isCurrentTurnTranscribing
import app.clawkietalkie.voice.reduce
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

// Port of test/drivingReducer.test.ts — pure-function tests for the Driving
// state machine.

class DrivingReducerTest {
    private val idle = initialDrivingContext
    private val recording = initialDrivingContext.copy(state = DrivingState.RECORDING)
    private val thinking = initialDrivingContext.copy(state = DrivingState.THINKING, lastUserText = "hello")
    private val ai = initialDrivingContext.copy(
        state = DrivingState.AI,
        lastUserText = "hello",
        lastReplyText = "hi there",
        liveReplyText = "hi there",
    )

    @Test
    fun `idle tap moves to recording and arms the mic`() {
        val (next, side) = reduce(idle, DrivingEvent.Tap())
        assertEquals(DrivingState.RECORDING, next.state)
        assertNull(next.error)
        assertEquals(listOf(DrivingSideEffect.START_MIC), side)
    }

    @Test
    fun `idle buffered ttsStart without text enters ai so playback can be silenced`() {
        val started = reduce(idle, DrivingEvent.TtsStart())
        assertEquals(DrivingState.AI, started.next.state)
        assertEquals("", started.next.lastReplyText)
        assertEquals("", started.next.liveReplyText)
        assertTrue(started.side.isEmpty())

        val silenced = reduce(started.next, DrivingEvent.Silence)
        assertEquals(DrivingState.IDLE, silenced.next.state)
        assertEquals(listOf(DrivingSideEffect.STOP_TTS), silenced.side)
    }

    @Test
    fun `idle ignores unrelated events`() {
        assertEquals(DrivingState.IDLE, reduce(idle, DrivingEvent.Silence).next.state)
        assertEquals(DrivingState.IDLE, reduce(idle, DrivingEvent.SttDone("x")).next.state)
        assertEquals(DrivingState.IDLE, reduce(idle, DrivingEvent.TtsDone).next.state)
    }

    @Test
    fun `recording tap stops the mic and enters thinking`() {
        val (next, side) = reduce(recording, DrivingEvent.Tap())
        assertEquals(DrivingState.THINKING, next.state)
        assertEquals(listOf(DrivingSideEffect.STOP_MIC), side)
    }

    @Test
    fun `recording sttError cancels back to idle with the reason surfaced`() {
        val (next, side) = reduce(recording, DrivingEvent.SttError("mic_denied"))
        assertEquals(DrivingState.IDLE, next.state)
        assertEquals("mic_denied", next.error)
        assertEquals(listOf(DrivingSideEffect.CANCEL_MIC), side)
    }

    @Test
    fun `thinking replyDone stays thinking hides reply text and arms tts`() {
        val (next, side) = reduce(thinking, DrivingEvent.ReplyDone("sup"))
        assertEquals(DrivingState.THINKING, next.state)
        assertEquals("sup", next.pendingReplyText)
        assertEquals("", next.lastReplyText)
        assertEquals("", next.liveReplyText)
        assertEquals(listOf(DrivingSideEffect.ARM_TTS), side)
    }

    @Test
    fun `thinking ttsStart with pending reply reveals the response and enters ai`() {
        val (next, side) = reduce(
            thinking.copy(pendingReplyText = "sup", lastReplyText = "previous audible reply"),
            DrivingEvent.TtsStart(),
        )
        assertEquals(DrivingState.AI, next.state)
        assertEquals("", next.pendingReplyText)
        assertEquals("sup", next.lastReplyText)
        assertEquals("sup", next.liveReplyText)
        assertTrue(side.isEmpty())
    }

    @Test
    fun `thinking stale ttsStart without pending reply does not show previous turn text`() {
        val (next, side) = reduce(
            thinking.copy(pendingReplyText = "", lastReplyText = "previous audible reply", liveReplyText = ""),
            DrivingEvent.TtsStart(),
        )
        assertEquals(DrivingState.THINKING, next.state)
        assertEquals("", next.pendingReplyText)
        assertEquals("previous audible reply", next.lastReplyText)
        assertEquals("", next.liveReplyText)
        assertTrue(side.isEmpty())
    }

    @Test
    fun `thinking replyDone then ttsError preserves the reply as the last ai turn`() {
        val ctx = reduce(thinking, DrivingEvent.ReplyDone("unheard reply")).next
        val (next, side) = reduce(
            ctx,
            DrivingEvent.TtsError("openclaw_infer_tts_failed: fetch timeout after 30000ms"),
        )
        assertEquals(DrivingState.IDLE, next.state)
        assertEquals("openclaw_infer_tts_failed: fetch timeout after 30000ms", next.error)
        assertEquals("", next.pendingReplyText)
        assertEquals("unheard reply", next.lastReplyText)
        assertEquals("", next.liveReplyText)
        assertTrue(side.isEmpty())
    }

    @Test
    fun `thinking replyDone then ttsDone returns idle without exposing the pending reply`() {
        val ctx = reduce(thinking, DrivingEvent.ReplyDone("unheard reply")).next
        val (next, side) = reduce(ctx, DrivingEvent.TtsDone)
        assertEquals(DrivingState.IDLE, next.state)
        assertEquals("", next.pendingReplyText)
        assertTrue(side.isEmpty())
    }

    @Test
    fun `thinking replyError returns idle with the reason`() {
        val (next, _) = reduce(thinking, DrivingEvent.ReplyError("openclaw_failed"))
        assertEquals(DrivingState.IDLE, next.state)
        assertEquals("openclaw_failed", next.error)
    }

    @Test
    fun `thinking sttDone records the final transcript`() {
        val (next, side) = reduce(thinking, DrivingEvent.SttDone("final words"))
        assertEquals(DrivingState.THINKING, next.state)
        assertEquals("final words", next.lastUserText)
        assertTrue(side.isEmpty())
    }

    @Test
    fun `thinking tap while transcribing is ignored`() {
        val (next, side) = reduce(thinking, DrivingEvent.Tap(currentTurnTranscribing = true))
        assertEquals(DrivingState.THINKING, next.state)
        assertTrue(side.isEmpty())
    }

    @Test
    fun `thinking tap after transcription cancels the reply`() {
        val (next, side) = reduce(thinking, DrivingEvent.Tap(currentTurnTranscribing = false))
        assertEquals(DrivingState.IDLE, next.state)
        assertEquals(listOf(DrivingSideEffect.CANCEL_REPLY), side)
    }

    @Test
    fun `ai tap silences playback`() {
        val (next, side) = reduce(ai, DrivingEvent.Tap())
        assertEquals(DrivingState.IDLE, next.state)
        assertEquals(listOf(DrivingSideEffect.STOP_TTS), side)
    }

    @Test
    fun `ai ttsDone completes the turn`() {
        val (next, side) = reduce(ai, DrivingEvent.TtsDone)
        assertEquals(DrivingState.IDLE, next.state)
        assertTrue(side.isEmpty())
        assertEquals("hi there", next.lastReplyText)
    }

    @Test
    fun `ai ttsError surfaces the reason and clears live text`() {
        val (next, _) = reduce(ai, DrivingEvent.TtsError("audio_unsupported"))
        assertEquals(DrivingState.IDLE, next.state)
        assertEquals("audio_unsupported", next.error)
        assertEquals("", next.liveReplyText)
    }

    @Test
    fun `session reset returns to the initial context`() {
        val (next, side) = reduce(ai, DrivingEvent.SessionReset)
        assertEquals(initialDrivingContext, next)
        assertTrue(side.isEmpty())
    }

    @Test
    fun `session replay runs events in order`() {
        val replay = DrivingEvent.SessionReplay(
            events = listOf(
                DrivingEvent.SttDone("question"),
                DrivingEvent.ReplyDone("answer"),
                DrivingEvent.TtsStart(),
            ),
        )
        val (next, side) = reduce(thinking, replay)
        assertEquals(DrivingState.AI, next.state)
        assertEquals("answer", next.lastReplyText)
        assertTrue(DrivingSideEffect.ARM_TTS in side)
    }

    @Test
    fun `terminal replay events win over a non-idle hydration`() {
        val replay = DrivingEvent.SessionReplay(
            events = listOf(
                DrivingEvent.SttDone("question"),
                DrivingEvent.ReplyDone("answer"),
                DrivingEvent.TtsStart(),
                DrivingEvent.TtsDone,
            ),
            hydration = app.clawkietalkie.voice.DrivingHydration(
                context = initialDrivingContext.copy(
                    state = DrivingState.AI,
                    lastReplyText = "answer",
                    liveReplyText = "answer",
                ),
                armTts = true,
            ),
        )
        val (next, side) = reduce(thinking, replay)
        assertEquals(DrivingState.IDLE, next.state)
        assertEquals("", next.pendingReplyText)
        assertEquals("", next.liveReplyText)
        assertFalse(DrivingSideEffect.ARM_TTS in side)
    }

    // -- helper coverage (drivingLoop.ts exports) ------------------------

    @Test
    fun `displayedCaptionText prefers liveReplyText in ai state`() {
        val ctx = DrivingContext(state = DrivingState.AI, liveReplyText = "reply")
        assertEquals("reply", displayedCaptionText(ctx, "live"))
        assertEquals("live", displayedCaptionText(ctx.copy(state = DrivingState.RECORDING), "live"))
    }

    @Test
    fun `isCurrentTurnTranscribing only while thinking with an active transcript`() {
        assertTrue(
            isCurrentTurnTranscribing(
                DrivingState.THINKING,
                CurrentTurnTranscript(active = true, sttDone = false, text = "x"),
            ),
        )
        assertFalse(
            isCurrentTurnTranscribing(
                DrivingState.THINKING,
                CurrentTurnTranscript(active = true, sttDone = true, text = "x"),
            ),
        )
        assertFalse(
            isCurrentTurnTranscribing(
                DrivingState.RECORDING,
                CurrentTurnTranscript(active = true, sttDone = false, text = "x"),
            ),
        )
    }

    @Test
    fun `composeTranscript joins finals and current interim`() {
        assertEquals("a b c", composeTranscript(listOf("a", "b"), " c "))
        assertEquals("a b", composeTranscript(listOf("a", "b"), "  "))
        assertEquals("c", composeTranscript(emptyList(), "c"))
    }
}
